plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/orthogonal/OrthogonalRoutingGenerator.java

```diff
@@ -222,49 +304,74 @@ private void createHyperEdgeSegments(final Iterable<LNode> nodes, final PortSide
      *            second hyperedge segments
+     * @return the number of critical dependencies that were added
      */
-    private void createDependency(final HyperEdgeSegment he1, final HyperEdgeSegment he2) {
+    private int createDependencyIfNecessary(final HyperEdgeSegment he1, final HyperEdgeSegment he2) {
         // check if at least one of the two nodes is just a straight line; those don't
         // create dependencies since they don't take up a slot
         if (Math.abs(he1.getStartCoordinate() - he1.getEndCoordinate()) < TOLERANCE
                 || Math.abs(he2.getStartCoordinate() - he2.getEndCoordinate()) < TOLERANCE) {
-            return;
+            
+            return 0;
         }
 
         // compare number of conflicts for both variants
         int conflicts1 = countConflicts(he1.getOutgoingConnectionCoordinates(), he2.getIncomingConnectionCoordinates());
         int conflicts2 = countConflicts(he2.getOutgoingConnectionCoordinates(), he1.getIncomingConnectionCoordinates());
-
-        // compare number of crossings for both variants
-        int crossings1 = countCrossings(
-                he1.getOutgoingConnectionCoordinates(), he2.getStartCoordinate(), he2.getEndCoordinate());
-        crossings1 += countCrossings(
-                he2.getIncomingConnectionCoordinates(), he1.getStartCoordinate(), he1.getEndCoordinate());
-        int crossings2 = countCrossings(
-                he2.getOutgoingConnectionCoordinates(), he1.getStartCoordinate(), he1.getEndCoordinate());
-        crossings2 += countCrossings(
-                he1.getIncomingConnectionCoordinates(), he2.getStartCoordinate(), he2.getEndCoordinate());
-
-        int depValue1 = CONFLICT_PENALTY * conflicts1 + crossings1;
-        int depValue2 = CONFLICT_PENALTY * conflicts2 + crossings2;
-
-        if (depValue1 < depValue2) {
-            // create dependency from first hypernode to second one
-            new HyperEdgeSegmentDependency(he1, he2, depValue2 - depValue1);
-        } else if (depValue1 > depValue2) {
-            // create dependency from second hypernode to first one
-            new HyperEdgeSegmentDependency(he2, he1, depValue1 - depValue2);
-        } else if (depValue1 > 0 && depValue2 > 0) {
-            // create two dependencies with zero weight
-            new HyperEdgeSegmentDependency(he1, he2, 0);
-            new HyperEdgeSegmentDependency(he2, he1, 0);
+        
+        boolean criticalConflictsDetected =
+                conflicts1 == CRITICAL_CONFLICTS_DETECTED || conflicts2 == CRITICAL_CONFLICTS_DETECTED;
+        int criticalDependencyCount = 0;
+        
+        if (criticalConflictsDetected) {
+            // Check which critical dependencies have to be added
+            if (conflicts1 == CRITICAL_CONFLICTS_DETECTED) {
+                // hyperedge 1 MUST NOT be left of hyperedge 2, since that would cause critical conflicts
+                new HyperEdgeSegmentDependency(DependencyType.CRITICAL, he2, he1, 0);
```

Comment:
It looks confustion that the constructor is called here but nothing happens to the instance (I saw it was like this before). I presume the dependency is added internally to the segments themselves. Maybe a static method would improve readability? `HyperEdgeSegmentDependency.createAndAdd(type, he2 ...)`?

Reply:
Excellent point. I'll refactor that.